transit.md

```diff
@@ -217,3 +272,108 @@ both sides have their own nonce.
+4. (the optional `priority` field on the hints should be ignore for now, as it is under-specified to the
+  point of being useless.)
+5. (Similar to point 4., an interface letting the user specify this might be provided.)
+
+The first two points are easy to implement: after a connection established, wait some time more to see if a
+"better" connection shows up. The waiting time should be roughly be derived from the time that was needed for
+the first one, but with some upper bound to not annoy users.
+
+The third point is a bit fuzzier: the line quickly becomes blurry when VPNs and corporate network setups are
+involved. For IPv4, simply check whether the connection was made over the external IP address or if it uses
+a prefix commonly used for local networks. For IPv6, an easy heuristic is to say that a connection where both
+sides' IP addresses have a longer prefix in common is preferable.
+
+### Potential pitfalls and other
+
+- The socket options that must be set are really subtle, for example read
+  [this excellent StackOverflow answer](https://stackoverflow.com/a/14388707/6094756) for more details.
+  In short, you need to check whether the operating system supports reusing ports and which options
+  must be set—on some systems, `SO_REUSEADDR` suffices, others provide `SO_REUSEPORT`. If the system doesn't
+  support it, fall back on the "easy" way and/or relay servers.
+- To prevent port hijacking, make sure that the port may not be reused by other processes. Some systems provide
+  special exclusivity options, while others enable them by default.
+- From the first moment a socket is bound to a port, that socket must be kept open. Simply binding and not using
+  it is okay. If you just bind one to get a free port and then close it immediately, another application may
+  (accidentally) hijack the port until you want to use it again.
+- Generally, all implementations should bind only IPv6 sockets (`AF_INET6`). The kernel will take care of
+  translating the packets to IPv4 if required. That way, no special support is required to handle both stacks.
+  Conversely, only binding to `127.0.0.1` means that the application will only support IPv4.
+
+## Future Extensions
+
+* WebSocket: usable by web browsers, not too hard to use by normal computers,
+  requires direct (or relayed) TCP connection
+* WebRTC: usable by web browsers, hard-but-technically-possible to use by
+  normal computers, provides NAT hole-punching for "free"
+* (web browsers cannot make direct TCP connections, so interop between
+  browsers and CLI clients will either require adding WebSocket to CLI, or a
+  relay that is capable of speaking/bridging both)
+* I2P: like Tor, but not capable of proxying to normal TCP hints.
+* ICE-mediated STUN/STUNT: NAT hole-punching, assisted somewhat by a server
+  that can tell you your external IP address and port. Maybe implemented as a
+  uTP stream (which is UDP based, and thus easier to get through NAT).
```

Comment:
proabably worth adding a point about Tor onion services, since that got deleted in a paragraph above. Perhaps:

* Tor Onion services: one side could run an Onion service, allowing a "direct" Tor connection without using any relay

Reply:
Oh, did I delete too much somewhere? Everything that the Python implementation does should be covered by the spec.